certbot-dovecot/certbot_dovecot/dovecotparser.py

```diff
@@ -0,0 +1,115 @@
+    """A class that parses dovecot configuration with pyparsing."""
+
+    # pyparsing automatically uses whitespaces as delimeters for tokens.
+    # Here, we're setting default whitespace characters to ` ` and `\t`,
+    # excluding `\n` since a new line is used as a delimeter for dovecot
+    # configuration itself
+    ParserElement.setDefaultWhitespaceChars('')
+
+    # This is printables excluding the following characters: # { } , =
+    allowed_characters = (
+        "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!\"$%&"
+        "\\'()*+-./:;<>?@[\\]^_`|~"
+    )
+
+    newline = LineEnd()
+    same_line_space = Optional(Word(' \t').leaveWhitespace())
+    space = Optional(White().leaveWhitespace())
+
+    # Defining comments
+    comment = space + Literal('#') + restOfLine
+
+    # Defining key and value pair
+    key = Word(allowed_characters) + same_line_space + (Literal('='))
+
+    value_group = OneOrMore(
+        same_line_space
+        + Optional(Literal(',') + same_line_space)
+        + Word(allowed_characters)
+    )
+
+    key_value_pair = (
+        space + key + value_group + same_line_space
+    )
+
+    # Defining includes
+    # Order matters. It's a first item match.
+    include = (
+        space
+        + (Literal('!include_try') | Literal('!include'))
+        + same_line_space
+        + Word(allowed_characters)
+        + same_line_space
+    )
+
+    # Defining a block
+    block = Forward()
+
+    item = Group(
+        comment
+        | block
+        | (key_value_pair + Optional(comment))
+        | (include + Optional(comment))
+    ) + newline.suppress()
+
+    block_title = Group(
+        space + Word(allowed_characters) + same_line_space
+        # Allow title with 2 words
+        + Optional(Word(allowed_characters) + same_line_space)
+        + Literal('{') + same_line_space + Optional(comment)
+        + newline.suppress()
+    )
+
+    right_brace = Group(space + Literal('}') + same_line_space + Optional(comment))
+
+    # The use of the OneOrMore(newline) here is actually not an assumption.
+    # Dovecot requires every { to be followed by a new line, and every } to
+    # be preceded and followed by a new line. Also, every key and value
+    # pair have to be delimited by a new line.
+    # Here a problem arises of picking whether { should always be followed
+    # by a new line, or } should always be preceded by one. We are making
+    # the choice to make sure all key value pairs in addition to { are
+    # followed by a new line. This should maintain that  } is also always
+    # preceded by a new line.
+    block_body = (Group(
+                        ZeroOrMore(item)
+                  ) + right_brace)
+
+    # Blocks contain of a list containing 3 lists. The first item is the
+    # block name, while the second item is a list of the block contents,
+    # and the third is the right brace and spaces surrounding it
+    block << OneOrMore(block_title + block_body)
```

Comment:
I think I'd still prefer that you parse blocks into a structure similar to the nginx parsing code ( `[ "block_name", [ "block", "contents" ] ]` rather than `[["block_name", " ", "{"], ["block", "contents"], ["}"]]`. This is mostly because we want to share as much code as possible between the nginx and dovecot parsers!

Reply:
If we do this, how will we be able to preserve spacing around the braces?

In other words, how will we be able to differentiate between `block name { ` and `block name {`, and `} ` and `}`?

Reply:
From your current implementation, you might have to preserve newlines. For instance, here's how our nginx parser parses an oddly spaced block (`#` used to denote EOL)
```
 block  {   #
    subblock   statement#
  }  #
```
into:
```
[
[ [' ', 'block', '  '],                 # block name, + spacing before left brace
  [ '   \n   ', 'subblock', '   ', 'statement' ] # spacing after left brace
  '\n  '                          # spacing before right brace
],
'  '                   # spacing after right brace
]
```

Reply:
Are you sure about this?

I understand how we want parsing to be as similar as possible, but the delimiters differ. Dovecot configuration delimiter is a new line.

It will require some special case handling to parse and dump blocks correctly if we don't include the new line following the braces with the block content itself, requiring us to also add the spacing preceding the `\n`.

If you're sure about this I'll give it a go.

Reply:
That's reasonable! I think we can make the difference in list-structures work nicely with the new parsing refactor.

Reply:
So, should we merge this now? :)